File: test/app_release_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:ten_of_a_kind_poker/services/app_release_service.dart';

void main() {
  group('AppRelease.fromJson', () {
    test('reads a well-formed payload', () {
      final AppRelease release = AppRelease.fromJson(<String, Object?>{
        'latestBuild': 84,
        'message': 'A new version is available.',
        'storeUrl': 'https://play.google.com/store/apps/details?id=com.toak',
      });
      expect(release.latestBuild, 84);
      expect(release.message, 'A new version is available.');
      expect(
        release.storeUrl,
        'https://play.google.com/store/apps/details?id=com.toak',
      );
    });

    test('a build number sent as a string still parses', () {
      expect(
        AppRelease.fromJson(<String, Object?>{'latestBuild': '84'}).latestBuild,
        84,
      );
    });

    test('malformed payloads degrade to nulls rather than throwing', () {
      for (final Object? raw in <Object?>[null, 'nonsense', 7, <int>[1, 2]]) {
        final AppRelease release = AppRelease.fromJson(raw);
        expect(release.latestBuild, isNull);
        expect(release.message, isNull);
        expect(release.storeUrl, isNull);
      }
    });

    test('non-positive and unparseable build numbers are discarded', () {
      for (final Object? bad in <Object?>[0, -1, 'v84', '', null]) {
        final AppRelease release =
            AppRelease.fromJson(<String, Object?>{'latestBuild': bad});
        expect(release.latestBuild, isNull, reason: 'expected $bad rejected');
      }
    });

    test('blank text reads as absent so the client falls back to its own copy',
        () {
      final AppRelease release = AppRelease.fromJson(<String, Object?>{
        'message': '   ',
        'storeUrl': '',
      });
      expect(release.message, isNull);
      expect(release.storeUrl, isNull);
    });
  });

  group('AppReleaseService', () {
    test('a build that cannot identify itself never reports out of date', () {
      // Tests run without --dart-define=APP_BUILD, which is exactly the state
      // of a bare `flutter run`. Such a client must stay quiet rather than
      // tell every developer they are on an old version.
      expect(AppReleaseService.currentBuild, 0);
      expect(
        AppReleaseService()
            .isNewerThanRunning(const AppRelease(latestBuild: 9999)),
        isFalse,
      );
    });

    test('a release at or below the running build is not an update', () {
      final AppReleaseService service = AppReleaseService();
      expect(service.isNewerThanRunning(const AppRelease()), isFalse);
      expect(
        service.isNewerThanRunning(const AppRelease(latestBuild: 1)),
        isFalse,
      );
    });
  });
}

File: backend/test/app_release.test.js
const assert = require('node:assert/strict');
const { test } = require('node:test');

const {
  _asHttpsUrl,
  _shapeAppRelease,
} = require('../src/routes/app_release');

test('a well-formed release document passes through intact', () => {
  assert.deepEqual(
    _shapeAppRelease({
      latestBuild: 84,
      message: 'A new version of X Poker is available.',
      storeUrl: 'https://play.google.com/store/apps/details?id=com.toak',
    }),
    {
      latestBuild: 84,
      message: 'A new version of X Poker is available.',
      storeUrl: 'https://play.google.com/store/apps/details?id=com.toak',
    },
  );
});

test('a build number typed as a string in the console still counts', () => {
  assert.equal(_shapeAppRelease({ latestBuild: '84' }).latestBuild, 84);
});

test('a missing or malformed document yields nulls, never a throw', () => {
  for (const input of [null, undefined, 'nonsense', 42, []]) {
    assert.deepEqual(_shapeAppRelease(input), {
      latestBuild: null,
      message: null,
      storeUrl: null,
    });
  }
});

test('junk build numbers are dropped rather than shipped to clients', () => {
  for (const bad of [0, -3, 1.5e300, 'v84', '', '  ', {}, null, NaN, Infinity]) {
    assert.equal(
      _shapeAppRelease({ latestBuild: bad }).latestBuild,
      null,
      `expected ${JSON.stringify(bad)} to be rejected`,
    );
  }
});

test('a float build number truncates instead of reaching the client as a float', () => {
  assert.equal(_shapeAppRelease({ latestBuild: 84.9 }).latestBuild, 84);
});

test('only https store links survive', () => {
  assert.equal(_asHttpsUrl('https://example.com/app'), 'https://example.com/app');
  for (const bad of [
    'http://example.com/app',
    'javascript:alert(1)',
    'market://details?id=com.toak',
    'intent://scan#Intent;scheme=zxing;end',
    'data:text/html,<script>alert(1)</script>',
    'not a url',
    '',
    null,
    17,
  ]) {
    assert.equal(_asHttpsUrl(bad), null, `expected ${String(bad)} to be rejected`);
  }
});

test('an over-long message is truncated, not rejected wholesale', () => {
  const shaped = _shapeAppRelease({ message: 'x'.repeat(1000) });
  assert.equal(shaped.message.length, 240);
});

test('a blank message reads as absent so the client shows its own copy', () => {
  assert.equal(_shapeAppRelease({ message: '   ' }).message, null);
});
